src/NServiceBus.Persistence.DynamoDB/Saga/SagaPersister.cs

```diff
@@ -24,19 +33,152 @@ public SagaPersister(SagaPersistenceConfiguration configuration, IAmazonDynamoDB
 
         public async Task<TSagaData> Get<TSagaData>(Guid sagaId, ISynchronizedStorageSession session, ContextBag context, CancellationToken cancellationToken = default) where TSagaData : class, IContainSagaData
         {
-            var getItemRequest = new GetItemRequest
+            if (configuration.UsePessimisticLocking)
+            {
+                return await ReadWithLock<TSagaData>(sagaId, context, session, cancellationToken).ConfigureAwait(false);
+            }
+            else
             {
-                ConsistentRead = true,
-                Key = new Dictionary<string, AttributeValue>
+                // Using optimistic concurrency control
+                var getItemRequest = new GetItemRequest
+                {
+                    ConsistentRead = true,
+                    Key = new Dictionary<string, AttributeValue>
                     {
                         { configuration.PartitionKeyName, new AttributeValue { S = $"SAGA#{sagaId}" } },
                         { configuration.SortKeyName, new AttributeValue { S = $"SAGA#{sagaId}" } }, //Sort key
                     },
-                TableName = configuration.TableName
-            };
+                    TableName = configuration.TableName
+                };
+
+                var response = await dynamoDbClient.GetItemAsync(getItemRequest, cancellationToken).ConfigureAwait(false);
+                return !response.IsItemSet ? default : Deserialize<TSagaData>(response.Item, context);
+            }
+        }
+
+        async Task<TSagaData> ReadWithLock<TSagaData>(Guid sagaId, ContextBag context,
+            ISynchronizedStorageSession synchronizedStorageSession, CancellationToken cancellationToken)
+            where TSagaData : class, IContainSagaData
+        {
+            using var timedTokenSource = new CancellationTokenSource(configuration.LeaseAcquistionTimeout);
+            using var sharedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, timedTokenSource.Token);
+            cancellationToken = sharedTokenSource.Token;
+
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    DateTimeOffset now = DateTimeOffset.UtcNow.Add(dynamoDbClient.Config.ClockOffset);
```

Comment:
@danielmarbach is that the correct approach? Is there some way we could test this?

Reply:
We do the same in the SQS transport. Testing we would probably require a good fake. I guess we can further improve this later